pennylane/beta/tapes/cv_param_shift.py

```diff
@@ -0,0 +1,458 @@
+                # the operator has a zero gradient.
+                best.append("0")
+                continue
+
+            # For parameter-shift compatible CV gates, we need to check both the
+            # intervening gates, and the type of the observable.
+            best_method = "A"
+
+            if any(not k.supports_heisenberg for k in ops_between):
+                # non-Gaussian operators present in-between the operation
+                # and the observable. Must fallback to numeric differentiation.
+                best_method = "F"
+
+            elif m.obs.ev_order == 2:
+
+                if m.return_type is qml.operation.Expectation:
+                    # If the observable is second order, we must use the second order
+                    # CV parameter shift rule
+                    best_method = "A2"
+
+                elif m.return_type is qml.operation.Variance:
+                    # we only support analytic variance gradients for
+                    # first order observables
+                    best_method = "F"
+
+            best.append(best_method)
+
+        if all(k == "0" for k in best):
+            # if the operation is independent of *all* observables
+            # in the circuit, the gradient will be 0
+            return "0"
+
+        if "F" in best:
+            # one non-analytic observable path makes the whole operation
+            # gradient method fallback to finite-difference
+            return "F"
+
+        if "A2" in best:
+            # one second order observable makes the whole operation gradient
+            # require the second order parameter-shift rule
+            return "A2"
+
+        return "A"
+
+    @staticmethod
+    def _transform_observable(obs, Z, device_wires):
+        """Apply a Gaussian linear transformation to each index of an observable.
+
+        Args:
+            obs (.Observable): observable to transform
+            Z (array[float]): Heisenberg picture representation of the linear transformation
+            device_wires (.Wires): wires on the device the transformed observable is to be
+                measured on
+
+        Returns:
+            .Observable: the transformed observable
+        """
+        # Get the Heisenber representation of the observable
+        # in the position/momentum basis. The returned matrix/vector
+        # will have been expanded to act on the entire device.
+        if obs.ev_order > 2:
+            raise NotImplementedError("Transforming observables of order > 2 not implemented.")
+
+        A = obs.heisenberg_obs(device_wires)
+
+        if A.ndim != obs.ev_order:
+            raise ValueError(
+                "Mismatch between the polynomial order of observable and its Heisenberg representation"
+            )
+
+        # transform the observable by the linear transformation Z
+        A = A @ Z
+
+        if A.ndim == 2:
+            A = A + A.T
+
+        return qml.PolyXP(A, wires=device_wires, do_queue=False)

```

Comment:
If `A` matches `qml.X`, `qml.P`, `qml.NumberOperator`, do we want to return those instead? Or is the chance of that happening quite small?

Reply:
What we could do: we could have `qml.PolyXP` simply return `qml.X`, `qml.P`, etc. itself depending on the argument it is instantiated with.

We do a similar thing with `TensorN` (it simply returns `NumberOperator` if it was instantiated on a single wire).

Reply:
yes, it's always bothered me that we use this clunky function when a simpler one might work in some cases

Reply:
Have added a todo here summing up this thread